### ISSUE TITLE
[Feature Request] - The possibility to parse feature files and split them into smaller parts.

### ISSUE THREAD

**Issue by netbrain** [temyers/cucumber-jvm-parallel-plugin#43]
Would be cool if there were some way to push the concurrency of the cucumber features even more. And one way to do this is to split up a feature file into smaller parts.

Take the following feature

``` gherkin
Feature: Upload photos

  As a user,
  I should be able to upload my photos

  Background:
    Given I have an account
    And I have a couple of photos on my device
    And I have logged into my account

  Scenario Outline: As a user I should be able to upload one or more photos
    When I select <num> photos for upload
    Then I expect <photos> photos to appear as uploaded
    And I expect <events> events to appear in the event view

  Examples:
    | num | photos | events |
    |   1 |      1 |      1 |
    |   4 |      4 |      1 |

  Scenario: As a user I should not be able to upload duplicate photos
    When I upload the same photo twice
    Then I expect 1 photos to appear as uploaded
    And I expect 1 events to appear in the event view
```

If i'm to run this scenario then this would take quite a while as it really is comprised of 3 test runs. 

- one for "As a user i should be able to upload one or more photos" with the 1, 1 ,1 example
- then one for the 4,4,1 example
- and lastly "As a user I should not be able to upload duplicate photos"

The way this could be improved was if this plugin could create three feature files from this single file where the feature file "1.feature" would contain: 

``` gherkin
Feature: Upload photos

  As a user,
  I should be able to upload my photos

  Background:
    Given I have an account
    And I have a couple of photos on my device
    And I have logged into my account

  Scenario Outline: As a user I should be able to upload one or more photos
    When I select <num> photos for upload
    Then I expect <photos> photos to appear as uploaded
    And I expect <events> events to appear in the event view

  Examples:
    | num | photos | events |
    |   1 |      1 |      1 |
```

Then the next "2.feature" file contains

``` gherkin
Feature: Upload photos

  As a user,
  I should be able to upload my photos

  Background:
    Given I have an account
    And I have a couple of photos on my device
    And I have logged into my account

  Scenario Outline: As a user I should be able to upload one or more photos
    When I select <num> photos for upload
    Then I expect <photos> photos to appear as uploaded
    And I expect <events> events to appear in the event view

  Examples:
    | num | photos | events |
    |   4 |      4 |      1 |
```

And then lastly "3.feature" contains 

``` gherkin
Feature: Upload photos

  As a user,
  I should be able to upload my photos

  Background:
    Given I have an account
    And I have a couple of photos on my device
    And I have logged into my account


  Scenario: As a user I should not be able to upload duplicate photos
    When I upload the same photo twice
    Then I expect 1 photos to appear as uploaded
    And I expect 1 events to appear in the event view
```

And then finally we'd use the existing functionality of the plugin to create dedicated test runners for each feature file.

WDYT? Too tall an order? or is it doable? Heck, maybe it would be better to implement this directly into cucumber.

**Comment by temyers** [comment]
It's a great idea, 

Everything is doable, but would probably be a hell of a lot of work.

Options:
* Superficially, parse the feature file and split per scenario (as you describe).  This would need a detailed model of the feature file and ability to both read and write them - building on the Gherkin parser itself.

* Create a more specific parallel Junit runner, which would partition on a per scenario basis.  This would probably be more appropriate in Cucumber JVM though (and would mean this plugin is no longer required)

The first option is probably easier, but the second would be the better solution IMO.

Or having smaller feature files to begin with, which might make them easier to read.

However, being devil's advocate....
Why are you wanting to do it?

If it's because a feature file contains lots of scenarios that take a long time, is there a question whether it would be more appropriate to test that functionality elsewhere?

**Comment by temyers** [comment]
Pull request welcome.... :)

**Comment by netbrain** [comment]
It's especially usefull for my use case where i have similar scenarios that i would like to group within a single feature file. But as each scenario takes quite a while to finish (uploading large files) it would be best if the scenarios weren't run sequentially.

It was just a thought I had while waiting for this time consuming scenario to finish :)

**Comment by sugatmankar** [comment]
@netbrain @temyers I have implemented this using gherking api. :) its quite easy with gherking.
With Gherking i parsed all scenarios and scenario outlines from a feature file and while parsing gherkin itself extracting each one as seperate test case like @netbrain explain in first post, so i just write in new file with unique tag like "@123456789" (random 10 digits).
So with the help of this i am able to run all seanarios parallaly (tagwise parallel i implemented in my local repo, soon going to merge it to my forked repo.). My final goal was to run all scenarios also outline parallely.

**Comment by tommywo** [comment]
if you have feature:
```cucumber
Feature: Upload photos

  As a user,
  I should be able to upload my photos

  Background:
    Given I have an account
    And I have a couple of photos on my device
    And I have logged into my account

  Scenario Outline: As a user I should be able to upload one or more photos
    When I select <num> photos for upload
    Then I expect <photos> photos to appear as uploaded
    And I expect <events> events to appear in the event view

  Examples:
    | num | photos | events |
    |   4 |      4 |      1 |
    |   2 |      4 |      1 |
```
you could run it parallely by specifying specific example line
so like feature1.feature:18 and feature1.feature:19

@sugatmankar so i think you could not generate changed and chucked feature files but run them using specific lines

**Comment by sugatmankar** [comment]
@tommywo yes we can create runner such way. I did this breaking for uploading feature files to Quality center and mark test with each unique generated test id on scenario, this is quit easy for maintenance of our test cases with HPQC.

**Comment by temyers** [comment]
@sugatmankar Cool!

Would it have an impact on reporting though?

**Comment by sugatmankar** [comment]
@temyers No it wont affect reporting, it will work as it is.

**Comment by sugatmankar** [comment]
[tagwiseOutlinewiseIssueRerun]( https://github.com/sugatmankar/cucumber-jvm-parallel-plugin/tree/tagwiseOutlinewiseIssueRerun) Branch has done implementation of outline parallelization.

**Comment by temyers** [comment]
Fixed in 2.0.0-SNAPSHOT
